pkg/controllers/pod_controller.go

```diff
@@ -0,0 +1,326 @@
+	ns := pod.Namespace
+	topologyKey := pod.Annotations[jobset.ExclusiveKey]
+
+	// Get node the leader pod is running on.
+	var node corev1.Node
+	if err := r.Get(ctx, types.NamespacedName{Name: nodeName, Namespace: ns}, &node); err != nil {
+		// We'll ignore not-found errors, since there is nothing we can do here.
+		// A node may not exist temporarily due to a maintenance event or other scenarios.
+		log.Error(err, fmt.Sprintf("getting node %s", nodeName))
+		return "", client.IgnoreNotFound(err)
+	}
+
+	// Get topology (e.g. node pool name) from node labels.
+	topology, exists := node.Labels[topologyKey]
+	if !exists {
+		return "", fmt.Errorf("node does not have topology label: %s", topology)
+	}
+	return topology, nil
+}
+
+func (r *PodReconciler) assignGatedPodsToTopology(ctx context.Context, podList *corev1.PodList, topologyKey, topologyValue string) error {
+	// log := ctrl.LoggerFrom(ctx)
+
+	// For every pod besides pod index 0, add a node selector for this topology and remove the scheduling gate.
+	// If we fail to update an individual pod, continue updating the remaining pods but return the error after
+	// to trigger reconciliation.
+	var (
+		finalErr error
+		wg       sync.WaitGroup
+	)
+
+	for _, p := range podList.Items {
+		pod := &p
+		if isLeaderPod(pod) {
+			continue
+		}
+		// Skip this pod if scheduling gate was already removed. This means the pod reconciler
+		// has already processed it and assigned it to the same topology as the leader pod.
+		gateIndex := schedulingGateIndex(pod)
+		if gateIndex == gateDoesNotExist {
+			continue
+		}
+
+		//log.Info(fmt.Sprintf("adding node selector to pod: %s", pod.Name))
+		addNodeSelector(pod, topologyKey, topologyValue)
+
+		//log.Info(fmt.Sprintf("removing scheduling gate from pod: %s", pod.Name))
+		removeSchedulingGate(pod, gateIndex)
+
+		//log.Info(fmt.Sprintf("updating pod: %s", pod.Name))
+
+		wg.Add(1)
+		go func(pd *corev1.Pod) {
+			pdCopy := pd
+			err := r.Update(ctx, pdCopy)
+			if err != nil {
+				finalErr = err
+			}
+			wg.Done()
+		}(pod)
+	}
+	wg.Wait()
+	return finalErr
+}
+
+// findLeaderPod iterates through a list of pods and returns a reference to the one
+// with a job completion index annotation of 0.
+func findLeaderPod(podList *corev1.PodList) *corev1.Pod {
+	var leaderPod *corev1.Pod
+	for i, p := range podList.Items {
+		if isLeaderPod(&p) {
+			leaderPod = &podList.Items[i]
+			break
+		}
+	}
+	return leaderPod
+}
+
+// listPodsForJobKey returns a list of pods owned by a specific job, using the
+// jobKey (SHA1 hash of the namespaced job name) label selector.
+func (r *PodReconciler) listPodsForJob(ctx context.Context, ns, jobKey string) (*corev1.PodList, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	labelSelector := labels.SelectorFromSet(map[string]string{
+		jobset.JobKey: jobKey,
+	})
+
+	var podList corev1.PodList
+	if err := r.List(ctx, &podList, client.InNamespace(ns), &client.ListOptions{LabelSelector: labelSelector}); err != nil {
```

Comment:
we should use the index:

`r.List(ctx, &podList, client.InNamespace(ns), client.MatchingFields{podOwnerKey: jobName})`

note we need to pass the jobName that pod belongs to (which is what we used to build the index)